### PR TITLE
Validate pool filesystem names

### PR DESCRIPTION
**Pull request by trgill**
Check stratis pool/filesystem names to make sure they don't contain:
    - NULL characters
    - Path directives - ../foo
    - Absolute paths /foo
    - . or ..
    - Name that exceeds 255 characters
    - Leading or trailing spaces
    - Control characters

Addresses issues #1271 and #1263

### CODE REVIEW

File: src/engine/strat_engine/names.rs [hunk 1]
```diff
@@ -183,3 +186,52 @@ pub fn format_backstore_ids(pool_uuid: PoolUuid, role: CacheRole) -> (DmNameBuf,
...
+    if name_path.is_absolute() {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains an absolute path : {}", name),
+        ));
+    }
+    if name == "." || name == ".." {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains . or .. : {}", name),
```
Comment (tasleson) [inline src/engine/strat_engine/names.rs]: `Name is . or .. :`

Comment (trgill) [reply src/engine/strat_engine/names.rs]: Ok - thanks.

File: src/engine/strat_engine/names.rs [hunk 2]
```diff
@@ -183,3 +186,89 @@ pub fn format_backstore_ids(pool_uuid: PoolUuid, role: CacheRole) -> (DmNameBuf,
...
+    if name.contains('\u{0}') {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains NULL characters: {}", name),
+        ));
+    }
+    if name_path.components().collect::<Vec<_>>().len() != 1 {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains forbidden characters: {}", name),
```
Comment (mulkieran) [inline src/engine/strat_engine/names.rs]:
```suggestion
            format!("Name is a path with 0 or more than 1 components: {}", name),
```

File: src/engine/strat_engine/names.rs [hunk 3]
```diff
@@ -183,3 +186,89 @@ pub fn format_backstore_ids(pool_uuid: PoolUuid, role: CacheRole) -> (DmNameBuf,
...
+        ));
+    }
+    if name_path.is_absolute() {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains an absolute path : {}", name),
```
Comment (mulkieran) [inline src/engine/strat_engine/names.rs]:
```suggestion
            format!("Name is an absolute path : {}", name),
```

File: src/engine/strat_engine/names.rs [hunk 4]
```diff
@@ -183,3 +186,89 @@ pub fn format_backstore_ids(pool_uuid: PoolUuid, role: CacheRole) -> (DmNameBuf,
...
+        ));
+    }
+
+    if name.len() != name.trim().len() {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name contains leading or trailing space : {}", name),
+        ));
+    }
+    if name.chars().filter(|c| c.is_control()).count() > 0 {
```
Comment (mulkieran) [inline src/engine/strat_engine/names.rs]:
```suggestion
    if name.chars().any(|c| c.is_control()) {
```

File: src/engine/strat_engine/names.rs [hunk 5]
```diff
@@ -183,3 +186,89 @@ pub fn format_backstore_ids(pool_uuid: PoolUuid, role: CacheRole) -> (DmNameBuf,
...
+    if name == "." || name == ".." {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name is . or .. : {}", name),
+        ));
+    }
+    if name.len() > 255 {
+        return Err(StratisError::Engine(
+            ErrorEnum::Invalid,
+            format!("Name too long : {}", name),
```
Comment (mulkieran) [inline src/engine/strat_engine/names.rs]:
```suggestion
            format!("Name has more than 255 characters : {}", name),
```

**Comment by trgill** [comment]
Opened final - https://github.com/stratis-storage/stratisd/pull/1283

**Pull request closed without merging**